packages/contracts/contracts/SmartInvoiceSplitEscrow.sol

```diff
@@ -0,0 +1,587 @@
+        require(_milestone < amounts.length, "invalid milestone");
+
+        uint256 balance = IERC20(token).balanceOf(address(this));
+
+        if (_remaining >= 1) {
+            uint256 amount = amounts[_milestone];
+            if (_remaining == 1 && amount < balance) {
+                amount = balance;
+            }
+            require(balance >= amount, "insufficient balance");
+
+            // milestone = milestone + 1;
+            milestoneReleased[_milestone] = true;
+            uint256 daoAmount = (amount * daoFee) / 10000;
+            uint256 providerAmount = amount - daoAmount;
+            IERC20(token).safeTransfer(dao, daoAmount);
+            IERC20(token).safeTransfer(provider, providerAmount);
+            released = released + amount;
+            unchecked {
+                _remaining = _remaining - 1;
+            }
+            emit Release(_milestone, amount);
+        } else {
+            require(balance > 0, "balance is 0");
+
+            uint256 daoAmount = (balance * daoFee) / 10000;
+            uint256 providerAmount = balance - daoAmount;
+            IERC20(token).safeTransfer(dao, daoAmount);
+            IERC20(token).safeTransfer(provider, providerAmount);
+            released = released + balance;
+            emit Release(_milestone, balance);
+        }
+    }
+
+    /**
+     * @dev Internal function to release funds from the contract to the provider.
+     */
+    function _release() internal {
+        // client transfers locker milestone funds to provider and dao
+        uint256 currentMilestone = amounts.length - 1;
+        for (uint256 i = 0; i < amounts.length; i++) {
+            if (milestoneReleased[i] == false) {
+                currentMilestone = i;
+                break;
+            }
+        }
+        _release(currentMilestone);
+    }
+
+    /**
+     * @dev External function to release funds from the contract to the provider.
+     * Uses the internal `_release` function to perform the actual release.
+     */
+    function release() external override nonReentrant {
+        return _release();
+    }
+
+    /** @dev External function to release funds from the contract to the provider.
+     * Uses the internal `_release` function to perform the actual release.
+     * @param _milestone The milestone to release funds to
+     */
+    function release(uint256 _milestone) external override nonReentrant {
+        // client transfers locker funds upto certain milestone to provider and dao
+        require(_milestone < amounts.length, "invalid milestone");
+        require(!milestoneReleased[_milestone], "milestone already released");
+        _release(_milestone);
+    }
+
+    /** @dev External function to release funds from the contract to the provider.
+     * Uses the internal `_release` function to perform the actual release.
+     * @param _milestones The milestones to release funds
+     */
+    function release(uint256[] calldata _milestones) external nonReentrant {
```

Comment:
This looks quite complicated to me. Would it be possible to separate the PR with the splitting and the release bit? I prefer when PRs tackle one feature and I'm not sure this is necessary for splitting. I would personally hold off for further refinement on the milestone release bits, but I think the splitting is important to get merged in sooner.